### DISCUSSION TITLE
Any benchmark? How many simultaneous connections/messages can this framework handle on a single core?

### DISCUSSION THREAD

**Discussion by PasqualePuzio** [reflex-dev/reflex#1973] [General]
I know this is a very complex question and it's really hard to give an accurate answer since it depends on a variety of factors including the complexity of the app and its state, but I was wondering if you have ever run a benchmark on this framework and what kind of performance and scalability we can expect.

I'm especially "concerned" with the the websocket mechanism, how many websockets can be open at the same time and how many messages the backend can handle per second.

**Comment by picklelo** [comment]
We don't have any official benchmarks - but under the hood we're using FastAPI for the websocket server which uses async calls to should scale up to many concurrent connections. For context, on some of our launch days we're able to handle 8000 unique users in a day on a small 2 core / 4GB machine, so the scalability is decent so far, but we are making more performance improvements.

**Reply by PasqualePuzio** [reply]
Thank you for the information. Is there any particular setup or configuration to apply in order to handle many (potentially thousands) concurrent connections and be able to use all the cores available?

**Reply by picklelo** [reply]
If you run your backend using the `--env prod` flag, we use a Gunicorn server that automatically uses all the available cores.

See this line:
https://github.com/reflex-dev/reflex/blob/b1bab1206d113e19d19754d8118c7448cc471cb3/reflex/utils/exec.py#L167

And we set the `num_workers` here:
https://github.com/reflex-dev/reflex/blob/b1bab1206d113e19d19754d8118c7448cc471cb3/reflex/utils/processes.py#L28

We could potentially expose this in the config if it would be useful.
